File: python/arrays_hashing/encodeDecodeString.py
class Solution:

    def encode(self, strs):
        res = ""

        for string in strs:
            res += "#"
            res += str(len(string))
            res += string

        return res

    def decode(self, s):
        ans = []
        current = ""
        # 4 n e e t # 4 c  o  d  e  #  4  l  o  v  e  #  3  y  o  u
        i = 1
        while i < len(s):
            curr_len = int(s[i])
            k = i + 1
            for j in range(0, curr_len):
                current += s[k]
                k = k + 1
            ans.append(current)
            i = k + 1
            current = ""

        return ans


def main():
    s = Solution()
    input_list = ["neet", "code", "love", "you"]
    res = s.encode(input_list)
    s.decode(res)


if __name__ == "__main__":
    main()

File: cpp/arrays_hashing/encodeDecodeString.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

string encode(vector<string>& strs) {
    string ans = "";

    for (int i = 0; i < strs.size(); i++) {
        ans += to_string(strs[i].size());
        ans += "#";
        for (int j = 0; j < strs[i].size(); j++) ans += strs[i][j];
    }
    return ans;
}

vector<string> decode(string s) {
    vector<string> ans;
    int i = 0;
    while (i < s.length()) {
        string slen = "";
        while (s[i] != '#') {
            slen += s[i];
            i++;
        }
        i++;
        int len = stoi(slen);
        string current = "";
        for (int j = 0; j < len; j++) {
            current += s[i];
            i++;
        }
        ans.push_back(current);
    }
    return ans;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    vector<string> input = { "1,23", "45,6", "7,8,9" };
    string output = encode(input);
    vector<string> out = decode(output);
    for (auto element : out) cout << element << endl;

    return 0;
}
